### PR TITLE
debuginfo: Fix race when extracting debug info files

### PR DESCRIPTION
**Pull request by javierhonduco**
fixes: https://github.com/parca-dev/parca-agent/issues/420

## Context
When generating profiles, we might need to extract and upload the debug information present in the binaries so we can symbolise the addresses with DWARF. We do this here:

https://github.com/parca-dev/parca-agent/blob/5f8f1525173e34125a101c374191e1bd0c031b16/pkg/profiler/profiler.go#L612-L622

When profiling many processes, we might have a race condition. For example, the same executables (libc) might show up in multiple processes, and we will attempt to upload them all. 

## Why is this a problem?

The explanation of it is in the code comment below. The result is that we end up with corrupted debug files, which this PR fixes!

It's also a performance issue, as we are repeating lots of work (extracting binaries, attempting to send them) that unnecesarily burn CPU cycles both in the server and in the agent.

## Test plan

Before this PR we generated some empty or corrupt files, now it's clean:

```
[javierhonduco@localhost-live parca]$ file  tmp/*/debuginfo | grep ELF | wc -l
336
[javierhonduco@localhost-live parca]$ file  tmp/*/debuginfo | grep -v ELF
[javierhonduco@localhost-live parca]$ echo $?
0
```

Simbolisation still works:

image

### CODE REVIEW

**Comment by javierhonduco** [comment]
Linter

File: pkg/debuginfo/debuginfo.go [hunk 1]
```diff
@@ -53,12 +57,49 @@ func New(logger log.Logger, client Client, tmp string) *DebugInfo {
...
+// and once all the writers are done, the debug file looks is an ELF
+// with the correct bytes.
+//
+// However I don't believe there's any guarantees on this, so the
+// files aren't getting corrupted most of the time by sheer luck.
+//
+// These two helpers make sure that we don't try to extract + upload
+// the same buildID concurrently.
+func (di *DebugInfo) alreadyUploading(buildID string) bool {
+	di.uploadingCacheMutex.Lock()
```
Comment (brancz) [inline pkg/debuginfo/debuginfo.go:85]: Do we really need these locking mechanisms? The `cache.Cache` looks to me as if it's safe for concurrent use.

**Comment by brancz** [comment]
Very nice find!

File: pkg/debuginfo/debuginfo.go (hunk 1 above)
Comment (javierhonduco) [reply pkg/debuginfo/debuginfo.go]: Good catch! https://github.com/goburrow/cache/issues/22#issuecomment-791784927

**Review by kakkoyun — APPROVED** [review]
Really great find! And amazing work to fix it.
Thanks a lot @javierhonduco!

**Pull request merged**